### PR TITLE
Close #18323: Refactor TitleMenu window to class

### LINKED ISSUE #18323: Refactor window to class: TileMenu
Refers to file: https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2-ui/windows/TileMenu.cpp
See #13737 for more details on how to accomplish the task.

### PR DESCRIPTION
**Pull request by rzhao271**
Fixes https://github.com/OpenRCT2/OpenRCT2/issues/18323

TODO:
- [x] Does `WindowTitleMenuInvalidate` translate to `OnUpdate`, or should it be `OnPrepareDraw`?
- [x] Quitting out of the track designer manager seems slower than normal. Confirm whether this change causes that regression.

### CODE REVIEW

**Comment by rzhao271** [comment]
There's a `_filterRect` variable I need to convert to a private field :sweat_smile:

File: src/openrct2-ui/windows/TitleMenu.cpp [hunk 1]
```diff
@@ -123,178 +63,227 @@ static void WindowTitleMenuScenarioselectCallback(const utf8* path)
...
-    const auto& customMenuItems = OpenRCT2::Scripting::CustomMenuItems;
-    size_t i = 0;
-    for (const auto& item : customMenuItems)
+    void OnMouseDown(WidgetIndex widgetIndex) override
     {
-        if (item.Kind == OpenRCT2::Scripting::CustomToolbarMenuItemKind::Toolbox)
+        if (widgetIndex == WIDX_GAME_TOOLS)
         {
-            if (i == index)
+            rct_widget* widget = &widgets[widgetIndex];
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/TitleMenu.cpp:177]: Move this closer to its usage, around line 210.

File: src/openrct2-ui/windows/TitleMenu.cpp [hunk 2]
```diff
@@ -123,178 +63,227 @@ static void WindowTitleMenuScenarioselectCallback(const utf8* path)
...
         }
     }
-}
 
-static void WindowTitleMenuCursor(
-    rct_window* w, WidgetIndex widgetIndex, const ScreenCoordsXY& screenCoords, CursorID* cursorId)
-{
-    gTooltipTimeout = 2000;
-}
+    CursorID OnCursor(WidgetIndex widgetIndex, const ScreenCoordsXY& screenCoords, CursorID cursorId) override
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/TitleMenu.cpp:255]:
```suggestion
    CursorID OnCursor(WidgetIndex, const ScreenCoordsXY&, CursorID cursorId) override
```
Unused parameters may be omitted.

File: src/openrct2-ui/windows/TitleMenu.cpp [hunk 3]
```diff
@@ -123,178 +63,227 @@ static void WindowTitleMenuScenarioselectCallback(const utf8* path)
...
-{
-    _filterRect = { w->windowPos.x, w->windowPos.y + UpdateButtonDims.height, w->windowPos.x + w->width - 1,
-                    w->windowPos.y + MenuButtonDims.height + UpdateButtonDims.height - 1 };
-    if (OpenRCT2::GetContext()->HasNewVersionInfo())
+    void OnPrepareDraw() override
     {
-        w->widgets[WIDX_NEW_VERSION].type = WindowWidgetType::Button;
-        _filterRect.Point1.y = w->windowPos.y;
+        _filterRect = { windowPos.x, windowPos.y + UpdateButtonDims.height, windowPos.x + width - 1,
+                        windowPos.y + MenuButtonDims.height + UpdateButtonDims.height - 1 };
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/TitleMenu.cpp:264]:
```suggestion
        _filterRect = { windowPos + ScreenCoordsXY { 0, UpdateButtonDims.height }, 
                        windowPos + ScreenCoordsXY{ width - 1, MenuButtonDims.height + UpdateButtonDims.height - 1 };
```

File: src/openrct2-ui/windows/TitleMenu.cpp [hunk 4]
```diff
@@ -123,178 +63,227 @@ static void WindowTitleMenuScenarioselectCallback(const utf8* path)
...
+    {
+        gfx_filter_rect(&dpi, _filterRect, FilterPaletteID::Palette51);
+        DrawWidgets(dpi);
+    }
+};
+
+/**
+ * Creates the window containing the menu buttons on the title screen.
+ *  rct2: 0x0066B5C0 (part of 0x0066B3E8)
+ */
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/TitleMenu.cpp:282]:
```suggestion
/**
 * Creates the window containing the menu buttons on the title screen.
 */
```

**Review by tupaschoal — APPROVED** [review]
I did some testing and it seems fine.

**Pull request merged**